File: 2022/6.js
const fs = require("fs")

fs.readFile("../../AoC-input/2022/6/input.txt", (err, input) => {
    if (err) throw err;
    let data = input.toString()
    let lines = data.split('\n')

    let index = 3
    lines.forEach(line => {
        let allDifferentCharacters = false
        while (allDifferentCharacters === false) {
            let isSomeDifferentCharacters = false
            let letters = []
            for (let i = 3; i > 0; i--) {
                letters.push(line[index - i])
                for (let j = 0; j < letters.length; j++) {
                    if (line[index - i + 1] === line[index - 3 + j]) {
                        isSomeDifferentCharacters = true
                        i = 0
                        break
                    }
                }
            }
            if (isSomeDifferentCharacters) {
                index++
            }
            else {
                allDifferentCharacters = true
            }
        }
    })

    console.log(index + 1)
})

fs.readFile("../../AoC-input/2022/6/input.txt", (err, input) => {
    if (err) throw err;
    let data = input.toString()
    let lines = data.split('\n')

    let index = 13
    lines.forEach(line => {
        let allDifferentCharacters = false
        while (allDifferentCharacters === false) {
            let isSomeDifferentCharacters = false
            let letters = []
            for (let i = 13; i > 0; i--) {
                letters.push(line[index - i])
                for (let j = 0; j < letters.length; j++) {
                    if (line[index - i + 1] === line[index - 13 + j]) {
                        isSomeDifferentCharacters = true
                        i = 0
                        break
                    }
                }
            }
            if (isSomeDifferentCharacters) {
                index++
            }
            else {
                allDifferentCharacters = true
            }
        }
    })

    console.log(index + 1)
})

File: 2025/src/bin/6.rs
use std::{env, fs};

fn part1(file: &str) -> i64 {
    let mut matrix: Vec<Vec<i64>> = vec![];
    let mut is_addition = vec![];
    for line in file.lines() {
        if line.contains("+") {
            for c in line.split_whitespace() {
                is_addition.push(c == "+");
            }
        }
        else {
            let mut row = vec![];
            for number in line.split_whitespace() {
                row.push(number.parse::<i64>().unwrap());
            }
            matrix.push(row);
        }
    }

    let mut sum = 0;
    for i in 0..is_addition.len() {
        if is_addition[i] {
            for j in 0..matrix.len() {
                sum += matrix[j][i];
            }
        }
        else {
            let mut local_sum = 1;
            for j in 0..matrix.len() {
                local_sum *= matrix[j][i];
            }
            sum += local_sum;
        }
    }
    sum
}

fn part2(file: &str) -> i64 {
    let mut char_input: Vec<Vec<char>> = vec![];
    let mut is_addition = vec![];
    for line in file.lines() {
        if line.contains("+") {
            for c in line.split_whitespace() {
                is_addition.push(c == "+");
            }
        }
        
        let mut row = vec![];
        for c in line.chars() {
            row.push(c);
        }
        char_input.push(row);
    }

    let mut matrix: Vec<Vec<i64>> = vec![];
    let mut index = 0;
    for i in 0..char_input[0].len() {
        let mut number = 0;
        for j in 0..char_input.len() - 1 {
            if char_input[j][i] != ' ' {
                number = number * 10 + char_input[j][i].to_digit(10).unwrap() as i64;
            }
        }
        
        if number == 0 {
            index += 1;
            matrix.push(vec![]);
            continue;
        }

        if matrix.len() <= index {
            matrix.push(vec![]);
        }
        matrix[index].push(number);
    }

    let mut sum = 0;
    for i in 0..matrix.len() {
        if is_addition[i] {
            for j in 0..matrix[i].len() {
                sum += matrix[i][j];
            }
        }
        else {
            let mut local_sum = 1;
            for j in 0..matrix[i].len() {
                local_sum *= matrix[i][j];
            }
            sum += local_sum;
        }
    }
    sum
}

fn main() {
    let filename = env::args().nth(1).unwrap_or_else(|| "../../AoC-input/2025/6/input.txt".to_string());
    let file = fs::read_to_string(&filename).expect("Failed to read file");
    
    println!("Part 1: {}", part1(&file));
    println!("Part 2: {}", part2(&file));
}
